backends/backend_openssl3.c

```diff
@@ -2472,3 +2472,627 @@ static void openssl_ecdsa_backend(void)
+		logger(LOGGER_WARN, "RSA E missing\n");
+		return -EINVAL;
+	}
+
+	bld = OSSL_PARAM_BLD_new();
+	logger_binary(LOGGER_DEBUG, data->e.buf, data->e.len, "e");
+	logger_binary(LOGGER_DEBUG, data->p.buf, data->p.len, "p");
+	logger_binary(LOGGER_DEBUG, data->q.buf, data->q.len, "q");
+
+	e = BN_bin2bn((const unsigned char *) data->e.buf, (int)data->e.len, e);
+	CKNULL(e, -ENOMEM);
+
+	p = BN_bin2bn((const unsigned char *) data->p.buf, (int)data->p.len, p);
+	CKNULL(p, -ENOMEM);
+	if (BN_is_zero(p))
+		BN_one(p);
+	q = BN_bin2bn((const unsigned char *)data->q.buf, (int)data->q.len, q);
+	CKNULL(q, -ENOMEM);
+	OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, e);
+	if (p != NULL) {
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR1, p);
+	}
+	if (q != NULL) {
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR2, q);
+	}
+	params = OSSL_PARAM_BLD_to_param(bld);
+	ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
+	EVP_PKEY_keygen_init(ctx);
+	ret = EVP_PKEY_CTX_set_params(ctx, params);
+	data->keygen_success = 1;
+
+	if(EVP_PKEY_generate(ctx, &pkey) <= 0)
+		data->keygen_success = 0;
+	if(BN_check_prime(p, NULL, NULL) <= 0)
+		data->keygen_success = 0;
+	if(BN_check_prime(q, NULL, NULL) <= 0)
+		data->keygen_success = 0;
+
+out:
+	if(params)
+		OSSL_PARAM_free(params);
+	if(bld)
+		OSSL_PARAM_BLD_free(bld);
+	if(ctx)
+		EVP_PKEY_CTX_free(ctx);
+	return ret;
+}
+
+int openssl_rsa_keygen_internal(struct buffer *ebuf, uint32_t modulus,
+					EVP_PKEY **outkey, struct buffer *nbuf,
+					struct buffer *dbuf, struct buffer *pbuf,
+					struct buffer *qbuf)
+{
+	BIGNUM *e = NULL;
+	EVP_PKEY_CTX *ctx = NULL;
+	OSSL_PARAM_BLD *bld = NULL;
+	EVP_PKEY *pkey = NULL;
+	size_t p_len = 0, q_len=0, egen_len=0, d_len=0,n_len=0;
+	unsigned int retry = 0;
+	int ret = 0;
+	unsigned char *n , *p , *q , *d,*egen;
+
+	if (!ebuf->len) {
+		unsigned int a;
+		uint8_t bitsset = 0;
+
+		/* WARNING Buffer must be at least 3 bytes in size ! */
+		CKINT(alloc_buf(sizeof(unsigned int), ebuf));
+
+		/* generate random odd e */
+		RAND_bytes(ebuf->buf, (int)ebuf->len);
+		/* make sure it is odd */
+		ebuf->buf[ebuf->len - 1] |= 1;
+
+		for (a = 0; a < ebuf->len - 2; a++)
+			bitsset |= ebuf->buf[a];
+
+		/* Make sure that value is >= 65537 */
+		if (!bitsset)
+			ebuf->buf[ebuf->len - 3] |= 1;
+	}
+
+	logger(LOGGER_DEBUG, "modulus: %u\n", modulus);
+	logger_binary(LOGGER_DEBUG, ebuf->buf, ebuf->len, "e");
+
+#if 1
+	e = BN_bin2bn((const unsigned char *)ebuf->buf, (int)ebuf->len, e);
+	CKNULL(e, -ENOMEM);
+#else
+	e = BN_new();
+	CKNULL(e, -ENOMEM);
+	if (1 != BN_set_word(e, 65537)) {
+		logger(LOGGER_WARN, "BN_set_word() failed");
+		ret = -EFAULT;
+		goto out;
+	}
+#endif
+	do {
+		ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
+		EVP_PKEY_keygen_init(ctx);
+		EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, (int)modulus);
+		EVP_PKEY_CTX_set1_rsa_keygen_pubexp(ctx, e);
+		ret = EVP_PKEY_keygen(ctx, &pkey);
+		retry++;
+	} while (ret != 1 && retry < 100);
```

Comment:
Can you explain the reason behind setting retry to 100?

Reply:
https://github.com/smuellerDD/acvpparser/blob/a6acb96a64a13810dc445b4a2fb9c88ae48caeb8/backends/backend_openssl.c#L1897
followed smuellerDD code